schema-layer/data-structures/generics.md

```diff
@@ -0,0 +1,109 @@
+
+These early implementations of multi-block data structures exposed several problems.
+
+  * They are lacking self-description. A consumer of a graph containing these structures would have to have logic on top of IPLD and vary the way it performs operations on there data structures.
+  * Implementations of these data structures cannot perform operations on each other. In other words, multi-block data structures have a hard time building on top of each other.
+
+Since there wasn’t a standardized way to describe these data structures we couldn’t build libraries for paths and selectors that seamlessly supported them.
+
+As we started designing this system several other requirements surfaced.
+
+* Transparent encryption envelopes on the read and write side.
+* Advanced `Link` types that can support some form of mutability and link to paths within other data structures.
+* Flexible multi-block binary types.
+
+## Requirements
+
+IPLD Generics cannot be implemented without:
+
+  * The IPLD Data-Model. There are codec agnostic but do require that the full data model be implemented by the codec.
+  * Reserved `_type` property.
+
+While the `_type` property is reserved in every map regardless of where it appears in a block, IPLD Generics **MUST** include the value of `_type` in the root property of a discrete block. Note: you can still embed multiple IPLD Generics in a single block using [inline blocks]().
+
+### Version 1
+
+The `_type` property is a string identifier. This identifier is used to lookup the implementation and if it cannot be found by the host environment any operation is expected to throw an exception. 
+
+Implementations MUST NOT fallback to *Layer 1* operations of the contents of the node.
+
+### Version 2
+
+The `_type` property is a `Map`.
+
+The map must contain the following properties.
+
+  * `name` must be a string identifier.
+  * `engine` must be one of the following:
+	* `”IPLD/Engine/WASM/0”`
+
+Each additional property describes the implementation of every operation.
+
+*TODO: define structure of pointers to WASM functions*
+
+## Operations
+
+This section describes only the operations for which there is currently an implementation. For a more exhaustive list of operations we may support in the future read [IPLD Multi-block Collections](https://github.com/ipld/specs/blob/master/schema-layer/data-structures/multiblock-collections.md).
+
+The term “leaf” is used often below. When an operation is performed on a node it may make any number of additional calls to operations on other data structures with many results being produced and passed into continuations. “leaf” responses are those produced by an engine that are from the original target operation and not from subsequent operations trigger by the original operation.
```

Comment:
This took me a some time to understand, due to the lengths of the sentences. What about splitting it into smaller ones. Here's a proposal, take this as an idea and make it read even nicer:

The term “leaf” is used often below. When an operation is performed on a node it may make any number of additional calls to operations on other data structures. Those will produce many results which are passed into continuations. “leaf” responses are those that are from the original target operation and not from subsequent operations trigger by the original operation.